ui/completion.py

```diff
@@ -6,47 +6,77 @@
 from LSP.plugin.core.types import basescope2languageid
 from LSP.plugin.core.typing import List, Optional
 
-from .types import CopilotPayloadCompletion
-from .utils import clamp, get_copilot_view_setting, reformat, set_copilot_view_setting
+from ..constants import PLAIN_TEXT_SYNTAX
+from ..types import CopilotPayloadCompletion
+from ..utils import clamp, get_copilot_view_setting, reformat, set_copilot_view_setting
 
 
 class ViewCompletionManager:
-    def __init__(self, view: sublime.View) -> None:
-        self.view = view
+    # ------------- #
+    # view settings #
+    # ------------- #
 
     @property
     def is_visible(self) -> bool:
         """Whether Copilot's completion popup is visible."""
         return get_copilot_view_setting(self.view, "is_visible", False)
 
+    @is_visible.setter
+    def is_visible(self, value: bool) -> None:
+        set_copilot_view_setting(self.view, "is_visible", value)
+
+    @property
+    def is_waiting(self) -> bool:
+        """Whether the view is waiting for Copilot's completion response."""
+        return get_copilot_view_setting(self.view, "is_waiting_completion", False)
+
+    @is_waiting.setter
+    def is_waiting(self, value: bool) -> None:
+        set_copilot_view_setting(self.view, "is_waiting_completion", value)
+
     @property
     def completions(self) -> List[CopilotPayloadCompletion]:
-        """All `completions` in the view."""
+        """All `completions` in the view. Note that this is a copy."""
         return get_copilot_view_setting(self.view, "completions", [])
 
+    @completions.setter
+    def completions(self, value: List[CopilotPayloadCompletion]) -> None:
+        set_copilot_view_setting(self.view, "completions", value)
+
     @property
     def completion_index(self) -> int:
         """The index of the current chosen completion."""
         return get_copilot_view_setting(self.view, "completion_index", 0)
 
+    @completion_index.setter
+    def completion_index(self, value: int) -> None:
+        """The index of the current chosen completion."""
+        do_clamp = not self.view.settings().get("auto_complete_cycle", False)
+        set_copilot_view_setting(
+            self.view,
+            "completion_index",
+            self._tidy_completion_index(value, do_clamp=do_clamp),
```

Comment:
It is outside of the scope of this PR, but since we are refactoring some stuff, maybe we don't need this `do_clamp` param. Instead, `do_clamp = not self.view.settings().get("auto_complete_cycle", False)` could be hardcoded in the `_tidy_completion_index`.  
To be honest, I would get rid of it at all, as it is the responsibility of this setter to handle the index

Reply:
Done in [cf74887](https://github.com/TheSecEng/LSP-copilot/pull/18/commits/cf748870347d6f835ce10102d5ad830db9b238ad)